### PR TITLE
🎉 Source Greenhouse: enhance stream support

### LINKED ISSUE #1385: Source Greenhouse Full Stream Coverage
Support the missing streams from Greenhouse: https://developers.greenhouse.io/harvest.html

Activity Feed
Approvals  
Education  
EEOC  
Email Templates 
Offices 
Prospect Pools  
Scheduled Interviews  
Tags  
Tracking Links  
User Permissions  
User Roles

### PR DESCRIPTION
**Pull request by itaseskii**
## What
Increase Greenhouse stream support

Closes https://github.com/airbytehq/airbyte/issues/1385

## How
*Describe the solution*

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-greenhouse/source_greenhouse/greenhouse.yaml [hunk 1]
```diff
@@ -35,6 +35,14 @@ definitions:
       page_token_option:
         inject_into: "path"
       url_base: "*ref(definitions.requester.url_base)"
+  no-pagination-retriever:
+    type: SimpleRetriever
```
Comment (girarda) [inline airbyte-integrations/connectors/source-greenhouse/source_greenhouse/greenhouse.yaml:39]: nit: the requester could be defined here so only the path needs to be defined in the stream definitions

File: airbyte-integrations/connectors/source-greenhouse/source_greenhouse/schemas/activity_feed.json [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+              "employee_id": {
+                "type": "string"
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+}
```
Comment (girarda) [inline airbyte-integrations/connectors/source-greenhouse/source_greenhouse/schemas/activity_feed.json:141]: nit: there should be a newline at the end of the files

**Review by girarda — COMMENTED** [review]
This is great @itaseskii! Thank you for your contribution!

Just a question on why the user_permissions schema is empty and small nits.

File: airbyte-integrations/connectors/source-greenhouse/source_greenhouse/greenhouse.yaml (hunk 1 above)
Comment (itaseskii) [reply airbyte-integrations/connectors/source-greenhouse/source_greenhouse/greenhouse.yaml:39]: Do you suggest I do something like this
```
 no-pagination-retriever:
    type: SimpleRetriever
    name: "{{ options['name'] }}"
    primary_key: "{{ options['primary_key'] }}"
    record_selector:
      $ref: "*ref(definitions.selector)"
    requester:
      $ref: "*ref(definitions.requester)"
    paginator:
      type: NoPagination

 tags_stream:
    $ref: "*ref(definitions.base_stream)"
    $options:
      name: "tags"
      path: "tags/candidate"
    retriever:
      $ref: "*ref(definitions.no-pagination-retriever)"

 user_permissions_stream:
    $options:
      name: "user_permissions"
    primary_key: "id"
    schema_loader:
      $ref: "*ref(definitions.schema_loader)"
    retriever:
      $ref: "*ref(definitions.no-pagination-retriever)"
      requester:
        path: "users/{{ stream_slice.parent_id }}/permissions/jobs"
      stream_slicer:
        type: SubstreamSlicer
        parent_stream_configs:
          - stream: "*ref(definitions.users_stream)"
            parent_key: "id"
            stream_slice_field: "parent_id"

```

Is this a valid configuration? can I override the requester in the user_permissions_stream without specifying a reference to the requester being overridden with  $ref: "*ref(definitions.requester)"? Is it automatically picked up from the retriever?

File: airbyte-integrations/connectors/source-greenhouse/source_greenhouse/schemas/activity_feed.json (hunk 2 above)
Comment (itaseskii) [reply airbyte-integrations/connectors/source-greenhouse/source_greenhouse/schemas/activity_feed.json]: added new lines to every stream schema being defined

**Comment by itaseskii** [comment]
> why is `airbyte-integrations/connectors/source-greenhouse/source_greenhouse/schemas/user_permissions.json` empty?

I must have reverted the changes along with some other stuff from my last commit. I have added it back.

File: airbyte-integrations/connectors/source-greenhouse/source_greenhouse/greenhouse.yaml (hunk 1 above)
Comment (girarda) [reply airbyte-integrations/connectors/source-greenhouse/source_greenhouse/greenhouse.yaml:39]: You still need to reference it if you overload it. However, you can define the `path` in a top-level `$options` block and it'll get propagated 


```
 no-pagination-retriever:
    type: SimpleRetriever
    name: "{{ options['name'] }}"
    primary_key: "{{ options['primary_key'] }}"
    record_selector:
      $ref: "*ref(definitions.selector)"
    requester:
      $ref: "*ref(definitions.requester)"
    paginator:
      type: NoPagination

 tags_stream:
    $ref: "*ref(definitions.base_stream)"
    $options:
      name: "tags"
      path: "tags/candidate"
    retriever:
      $ref: "*ref(definitions.no-pagination-retriever)"

 user_permissions_stream:
    $options:
      name: "user_permissions"
      path: "users/{{ stream_slice.parent_id }}/permissions/jobs"
    primary_key: "id"
    schema_loader:
      $ref: "*ref(definitions.schema_loader)"
    retriever:
      $ref: "*ref(definitions.no-pagination-retriever)"
      stream_slicer:
        type: SubstreamSlicer
        parent_stream_configs:
          - stream: "*ref(definitions.users_stream)"
            parent_key: "id"
            stream_slice_field: "parent_id"
```

**Review by girarda — APPROVED** [review]
This is great! Thank you for your contribution @itaseskii ! @marcosmarxm can you or someone from CAT help merging this into master?

**Pull request closed without merging**